pkg/frontend/transport/server_timing_middleware.go

```diff
@@ -0,0 +1,68 @@
+package transport
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	querier_stats "github.com/cortexproject/cortex/pkg/querier/stats"
+
+	"github.com/felixge/httpsnoop"
+)
+
+const HeaderName = "Server-Timing"
+
+func ServerTimingMiddleware(next http.Handler, statsEnabled bool) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		if !statsEnabled {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		var (
+			headerWritten = false
+			startTime     = time.Now()
+		)
+
+		hooks := httpsnoop.Hooks{
+			WriteHeader: func(next httpsnoop.WriteHeaderFunc) httpsnoop.WriteHeaderFunc {
+				return func(code int) {
+					writeTimingHeader(startTime, &headerWritten, w.Header(), r)
+					next(code)
+				}
+			},
+			Write: func(next httpsnoop.WriteFunc) httpsnoop.WriteFunc {
+				return func(p []byte) (int, error) {
+					n, err := next(p)
+					writeTimingHeader(startTime, &headerWritten, w.Header(), r)
+					return n, err
+				}
+			},
+		}
+
+		w = httpsnoop.Wrap(w, hooks)
+		next.ServeHTTP(w, r)
+
+		writeTimingHeader(startTime, &headerWritten, w.Header(), r)
+	})
+}
+
+func writeTimingHeader(startTime time.Time, headerWritten *bool, headers http.Header, r *http.Request) {
+	if !*headerWritten {
+		parts := make([]string, 0)
+		stats := querier_stats.FromContext(r.Context())
+		if stats != nil {
+			parts = append(parts, statsValue("querier_wall_time", stats.WallTime))
```

Comment:
What's the expected unit for the duration in the `Server-Timing` specs? I'm asking cause we return nanoseconds and I'm wondering how it gets displayed by the browser dev tools.

Reply:
Also, we should never directly read `stats.WallTime` but use `stats.LoadWallTime()`.